src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs

```diff
@@ -235,40 +233,231 @@ internal void WriteTopLevelProperty(ODataProperty property)
             IDuplicatePropertyNameChecker duplicatePropertyNameChecker,
             ODataResourceMetadataBuilder metadataBuilder)
         {
-            WriterValidationUtils.ValidatePropertyNotNull(propertyInfo);
+            ValidatePropertyInfo(propertyInfo, owningType, isTopLevel, duplicatePropertyNameChecker);
 
-            string propertyName = propertyInfo.Name;
+            if (currentPropertyInfo.MetadataType.IsUndeclaredProperty)
+            {
+                WriteODataTypeAnnotation(propertyInfo, isTopLevel);
+            }
 
-            if (this.JsonLightOutputContext.MessageWriterSettings.Validations != ValidationKinds.None)
+            WriteInstanceAnnotation(propertyInfo, isTopLevel, currentPropertyInfo.MetadataType.IsUndeclaredProperty);
+
+            ODataStreamPropertyInfo streamInfo = propertyInfo as ODataStreamPropertyInfo;
+            if (streamInfo != null && !(this.JsonLightOutputContext.MetadataLevel is JsonNoMetadataLevel))
             {
-                WriterValidationUtils.ValidatePropertyName(propertyName);
+                Debug.Assert(!isTopLevel, "Stream properties are not allowed at the top level.");
+                WriteStreamValue(streamInfo, propertyInfo.Name, metadataBuilder);
             }
+        }
 
-            if (!this.JsonLightOutputContext.PropertyCacheHandler.InResourceSetScope())
+        /// <summary>
+        /// Asynchronously writes an <see cref="ODataProperty" /> to the given stream. This method creates an
+        /// async buffered stream and writes the property to it.
+        /// </summary>
+        /// <param name="property">The property to write.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal Task WriteTopLevelPropertyAsync(ODataProperty property)
+        {
+            Debug.Assert(property != null, "property != null");
+            Debug.Assert(!(property.Value is ODataStreamReferenceValue), "!(property.Value is ODataStreamReferenceValue)");
+
+            return this.WriteTopLevelPayloadAsync(
+                async () =>
+                {
+                    await this.AsynchronousJsonWriter.StartObjectScopeAsync().ConfigureAwait(false);

```

Comment:
@marabooy If you notice, we're eliding the call to `WriteTopLevelPayloadAsync` - since its the only _material_ statement within `WriteTopLevelPropertyAsync` method (i.e. the `WriteTopLevelPropertyAsync` is a pass-through). As for the delegate, we can't elide the await since we have multiple statements that should be executed in chronological order.
[Here](https://github.com/OData/odata.net/pull/2030/files#diff-3b9d6424a2dfe7cd670aec4af41eb01d3962575eef4efe31df83c73562ad0011R352) is another call to `WriteTopLevelPayloadAsync` and you'll notice that because there's only one method called when the delegate is invoked, we're actually eliding the await.